src/bijectors/leaky_relu.jl

```diff
@@ -0,0 +1,93 @@
+"""
+    LeakyReLU{T, N}(α::T) <: Bijector{N}
+
+Defines the invertible mapping
+
+    x ↦ x if x ≥ 0 else αx
+
+where α > 0.
+"""
+struct LeakyReLU{T, N} <: Bijector{N}
+    α::T
+end
+
+LeakyReLU(α::T; dim::Val{N} = Val(0)) where {T<:Real, N} = LeakyReLU{T, N}(α)
+LeakyReLU(α::T; dim::Val{N} = Val(D)) where {D, T<:AbstractArray{<:Real, D}, N} = LeakyReLU{T, N}(α)
+
+up1(b::LeakyReLU{T, N}) where {T, N} = LeakyReLU{T, N + 1}(b.α)
+
+# (N=0) Univariate case
+function (b::LeakyReLU{<:Any, 0})(x::Real)
+    mask = x < zero(x)
+    return mask * b.α * x + !mask * x
+end
+(b::LeakyReLU{<:Any, 0})(x::AbstractVector{<:Real}) = map(b, x)
+
+function Base.inv(b::LeakyReLU{<:Any,N}) where N
+    invα = inv.(b.α)
+    return LeakyReLU{typeof(invα),N}(invα)
+end
+
+function logabsdetjac(b::LeakyReLU{<:Any, 0}, x::Real)
+    mask = x < zero(x)
+    J = mask * b.α + (1 - mask) * one(x)
+    return log(abs(J))
+end
+logabsdetjac(b::LeakyReLU{<:Real, 0}, x::AbstractVector{<:Real}) = map(x -> logabsdetjac(b, x), x)
+
+
+# We implement `forward` by hand since we can re-use the computation of
+# the Jacobian of the transformation. This will lead to faster sampling
+# when using `rand` on a `TransformedDistribution` making use of `LeakyReLU`.
+function forward(b::LeakyReLU{<:Any, 0}, x::Real)
+    mask = x < zero(x)
+    J = mask * b.α + !mask * one(x)
+    return (rv=J * x, logabsdetjac=log(abs(J)))
+end
+
+# Batched version
+function forward(b::LeakyReLU{<:Any, 0}, x::AbstractVector)
+    J = let z = zero(x), o = one(x)
```

Comment:
IMO that's not a good fix since it destroys the whole point of just having one broadcast expression that is fused together. Now we would allocate two additional vectors of ones and zeros just for Tracker.

Did you check if we can work around the Tracker bug by using a separate function `ispositive(x) = x > zero(x)` and `isnegative(x) = x < zero(x)` that we can include in the broadcast expression instead of the explicit `x < zero(x)` etc statements?

Reply:
> Now we would allocate two additional vectors of ones and zeros just for Tracker.

Not two vectors; just two _numbers_, right? (I realize the above impl was wrong, it was supposed to contain `eltype`)

Reply:
Also, even if we use `isnegative`, we're running into the same issue because we're broadcasting over `one(x)`, so we would have to also have this in a separate function, e.g. `mul1(x, y) = x * one(y)` or something :confused:

Reply:
With `eltype` it's just a number - IMO it's still unfortunate since there would be no need for this type-based computation if Tracker would be fixed (I opened a PR: https://github.com/FluxML/Tracker.jl/pull/85). In general it's better to not use types if not needed since the instances contain more information (similar argument for why generated functions should be used only if needed). It *should* just work with this single broadcast expression here.

Apart from that, are the `let` blocks actually needed?

Reply:
BTW it seems the tests still fail?

Reply:
It still breaks, but now the only one failing is `LeakyReLU{..., 1}`. Again it hits a snag at `materlialize`:
```julia
Variables
  b::LeakyReLU{Float64,1}
  x::TrackedArray{…,Array{Float64,2}}
  z::Tracker.TrackedReal{Float64}

Body::Any
1 ─ %1 = Bijectors.eltype(x)::Core.Compiler.Const(Tracker.TrackedReal{Float64}, false)
│        (z = Bijectors.zero(%1))
│   %3 = Base.broadcasted(Bijectors.:<, x, z)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(<),Tuple{TrackedArray{…,Array{Float64,2}},Tracker.TrackedReal{Float64}}}
│   %4 = Base.getproperty(b, :α)::Float64
│   %5 = Base.broadcasted(Bijectors.:*, %3, %4)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(<),Tuple{TrackedArray{…,Array{Float64,2}},Tracker.TrackedReal{Float64}}},Float64}}
│   %6 = Base.broadcasted(Bijectors.:>, x, z)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(>),Tuple{TrackedArray{…,Array{Float64,2}},Tracker.TrackedReal{Float64}}}
│   %7 = Base.broadcasted(Bijectors.:*, %6, x)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(>),Tuple{TrackedArray{…,Array{Float64,2}},Tracker.TrackedReal{Float64}}},TrackedArray{…,Array{Float64,2}}}}
│   %8 = Base.broadcasted(Bijectors.:+, %5, %7)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(+),Tuple{Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(<),Tuple{TrackedArray{…,Array{Float64,2}},Tracker.TrackedReal{Float64}}},Float64}},Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(>),Tuple{TrackedArray{…,Array{Float64,2}},Tracker.TrackedReal{Float64}}},TrackedArray{…,Array{Float64,2}}}}}}
│   %9 = Base.materialize(%8)::Any
└──      return %9
```

Reply:
> IMO it's still unfortunate since there would be no need for this type-based computation if Tracker would be fixed (I opened a PR: FluxML/Tracker.jl#85)

100% agree! And nice! 

> Apart from that, are the let blocks actually needed?

Nah, I just added them to make explicit that it's only needed for this particular statement, and not to clutter the names in the full function. Was nicer IMO, but ofc subjective.